perllib/FixMyStreet/Cobrand/Bromley.pm

```diff
@@ -1146,6 +1153,255 @@ sub garden_waste_cost {
     return $self->feature('payment_gateway')->{ggw_cost} * $bin_count;
 }
 
+sub waste_payment_type {
+    my ($self, $type, $ref) = @_;
+
+    my ($sub_type, $category);
+    if ( $type eq 'Payment: 01' ) {
+        $category = 'Garden Subscription';
+        $sub_type = $self->waste_subscription_types->{New};
+    } elsif ( $type eq 'Payment: 17' ) {
+        $category = 'Garden Subscription';
+        if ( $ref ) {
+            $sub_type = $self->waste_subscription_types->{Amend};
+        } else {
+            $sub_type = $self->waste_subscription_types->{Renew};
+        }
+    } elsif ( $type eq 'AUDDIS: 0C' ) {
+        $sub_type = 0;
+        $category = 'Cancel';
+    }
+
+    return ($category, $sub_type);
+}
+
+sub waste_reconcile_direct_debits {
+    my $self = shift;
+
+    my $today = DateTime->now;
+    my $start = $today->clone->add( days => -4 );
+
+    my $config = $self->feature('payment_gateway');
+    my $i = Integrations::Pay360->new({
+        config => $config
+    });
+
+    my $recent = $i->get_recent_payments(
+        start => $start,
+        end => $today
+    );
+
+
+    RECORD: for my $payment ( @$recent ) {
+
+        my $date = $payment->{CollectionDate};
+        my ($category, $type) = $self->waste_payment_type ( $payment->{Type}, $payment->{YourRef} );
+
+        next unless $category && $date;
+
+        my $payer = $payment->{PayerReference};
+        my $reference = $payer;
+
+        (my $uprn = $payer) =~ s/^GGW//;
+
+        my $len = length($uprn);
+        my $rs = FixMyStreet::DB->resultset('Problem')->search({
+            extra => { like => '%uprn,T5:value,I' . $len . ':'. $uprn . '%' },
+        },
+        {
+                order_by => { -desc => 'created' }
+        });
+
+        my $handled;
+
+        if ( $type && $type eq $self->waste_subscription_types->{Renew} ) {
```

Comment:
This is quite a big block; I see the commit message, but might be worth a bit more explanation? It's not that easy to follow. So currently, if I've understood it right:
1. If the payment is a renew, skip if we find a Renew report with the same date, but if we find a New report, take the oldest one (should that be the newest one? Or I guess will they all be identical anyway...), and create a copy of it as a Renew report. (That copy means next time it'll see that and skip)
2. If the payment is a processed cancel, skip if we find a Cancel report (why is this a different category to the others?) with the same date, if we find a New report (again, the oldest one?), create a duplicate Cancel report.
3. Otherwise, if a paid payment, fetch all reports (except cancel ones), skip any if reference/date matches, or a New report where date matches, otherwise set payerReference if not set, set dd_date, and confirm the report (whatever it is).

Not sure, but would it perhaps be simpler to do one loop of the matching reports? Dunno, just feels like it could be made more straightforward somehow.

Reply:
Cancel is a different category as it is a different event code in Echo.